### PR TITLE
Refactor Intercom Service

### PR DESCRIPTION
**Pull request by jalcine-cfa**
- Replaces some logic around how calls to Intercom are done.
- Sets a stricter signature when making said calls.

### CODE REVIEW

File: app/controllers/mailgun_webhooks_controller.rb [hunk 1]
```diff
@@ -106,7 +106,7 @@ def create_incoming_email
             client: contact_record.client,
             body: contact_record.body,
             email_address: contact_record.sender,
-            has_documents: contact_record.attachment_count&.nonzero?
+            has_documents: (contact_record&.attachment_count || 0).nonzero?
```
Comment (alaroia-cfa) [inline app/controllers/mailgun_webhooks_controller.rb:109]: Looks like we can drop the first `&` thankfully!

File: spec/controllers/mailgun_webhooks_controller_spec.rb [hunk 2]
```diff
@@ -93,8 +93,10 @@
             end.to change(IncomingEmail, :count).by(0).and change(Client, :count).by(0)
             expect(IntercomService).to have_received(:create_message).with(
               email_address: sender_email,
-              inform_of_handoff: false,
-              body: "Hi Alice,\n\nThis is Bob.\n\nI also attached a file."
+              body: "Hi Alice,\n\nThis is Bob.\n\nI also attached a file.",
```
Comment (alaroia-cfa) [inline spec/controllers/mailgun_webhooks_controller_spec.rb:96]: The text here says that Bob attached a file but we're not doing `has_documents: true`. Let's see if we should fix the copy here.

**Review by alaroia-cfa — COMMENTED** [review]
Some comments on this commit; will then look at the PR as a whole

File: app/controllers/mailgun_webhooks_controller.rb [hunk 3]
```diff
@@ -22,10 +22,12 @@ def create_incoming_email
...
 
-        IntercomService.create_intercom_message(
+        IntercomService.create_message(
+          client: nil,
+          phone_number: nil,
           email_address: sender_email,
-          inform_of_handoff: false,
-          body: params["stripped-text"] || params["body-plain"]
+          body: params["stripped-text"] || params["body-plain"],
+          has_documents: false
```
Comment (alaroia-cfa) [inline app/controllers/mailgun_webhooks_controller.rb:30]: Arguably it'd be nice if we validated that `has_documents` is `true` when it should be.

File: app/services/intercom_service.rb [hunk 4]
```diff
@@ -132,24 +111,27 @@ def self.intercom_contact_attributes(phone_number: nil, email_address: nil, clie
...
-
-  def self.reply_to_existing_intercom_thread(contact_id, body)
-    intercom.conversations.reply_to_last(
-      intercom_user_id: contact_id,
-      type: 'user',
-      message_type: 'comment',
-      body: body
-    )
+  def self.create_new_intercom_thread(contact, body)
+    # Read https://web.archive.org/web/20230118155416/https://forum.intercom.com/s/question/0D55c00005vHaPsCAK/create-conversation-for-lead-throws-an-error-user-not-found
```
Comment (alaroia-cfa) [inline app/services/intercom_service.rb:115]: Bah I tested that archive.org link and it's blank. Maybe archive.is is a better choice.

https://developers.intercom.com/intercom-api-reference/reference/create-a-conversation says we should `lead` or `user`. The forum thread says we should use `user`. If we reference the forum thread, we have to change the code here.

I'll come back to this after we do our manual testing.

Comment (alaroia-cfa) [reply app/services/intercom_service.rb:115]: archive.is links fwiw

https://archive.is/4oUOL aka https://developers.intercom.com/intercom-api-reference/reference/create-a-conversation

https://archive.is/7wTQQ aka https://forum.intercom.com/s/question/0D55c00005vHaPsCAK/create-conversation-for-lead-throws-an-error-user-not-found

File: config/credentials/development.yml.enc [hunk 5]
```diff
@@ -1 +0,0 @@
...IRtwshyIILW132iaXQsftrddad5E8r4zeo/HsfJygiQnjT83kPukF3COUTMyUIAjt4gXCLeO622ugAulaTRtaeK0ItMIhl0Jlo+MbZCc4eJlx7UFA+snMii8K5Q/7tmBSQj2fi5xo81m1S8vV0iLwUDInfjhiDnbDKFiFZ1rB+He51mNuhfwCLdB/Q01h65iKTXxTGbJub5800qi0RkGK3HRWKhM+HrhFn4S+WLWy1BSNHpLrkA0Y4qMwzFDXpOaoJymH3exvlg97F7Tj4QnUGDw3ZhP6J22Hnwcq3op8ISvAy8LW2RmS7aAQiWUzlK1ZMy/8aeTq3DBdMdTMM4Neqvppa9/Mfad/yia4BGRvJRPhqSIZN4Tzzd8Gwmcgrw85CgY4mQt2yKoPckF8hTnMv1fWUeDIo+7XvqbFUl5xgQbSQ30Oap3wtCb5a3lowdvajGqe2rJWIMGORVxpZrZ2zjV0eSuu4OQeECCIpHbCYOFgVNrsmD/Oxl8OLPGigrw55IFrOwq4+4V64fcgf5pzoMmy8TIDlKYlC/Kj0D0CkMfQXTYymK2WCPn3iSj1e7OvKU4nYFi7osm0iFOlzehC2mWsH6EiXknznFbjoN4TOTPqeyoT6OvW8EL8CXXcHdGTYI22IOD3WN2cCVE6jLpGZZqNF40vICj98XkzxFQ79UNXirklr60G2oYeTLitLpXjcF5dfmYZiP0hPfHK0rajuI1MFPnKPSGgyA50wPt3xEgn86kJSEz056H62EfCBzfh9+BcvjzOV+Vw/wKGXV7b78F3gjbyba0X8QPMqAkb8HnDjJ+2tR1nuBKWumiIlyAqp9sX9lIpYpFdKPBUZi/YV8R/Ik0k0lBgbPeSYGjZVxGHnnEB0an2jHaFHHzecKtUDJwG5csyCL0e2xzGmSZawBu2YlpFVykGOSNIJ6wlgoas/Lv8o+EOAkLAR8oc9vYCvhtM9akwkc8y7W5dazLKY3pWnPUvsO8COGgHu4AB9BtFmc24kr4jbpDvpZVOA+cGIc9hbwF88vaZk1HGFqwhTWXkRx5SiMR7l3/NVC3JBuOQWNx0oSVs3tjqHT3j3bfWLttwXSoIaUruY5VQ7cEEl484kXllIqGu+Yhn2fUBWaFZZow38ZlahTTUiRLp9ee1A==--x6r5Mlal38NTHMeo--kRq2lfhdfSXaSJwLGXUojg==
```
Comment (alaroia-cfa) [inline config/credentials/development.yml.enc:1]: Let's revert this change from this PR

**Review by alaroia-cfa — COMMENTED** [review]
Some ideas here; I'll do the manual testing now before making these code changes.

File: app/controllers/mailgun_webhooks_controller.rb (hunk 3 above)
Comment (jalcine-cfa) [reply app/controllers/mailgun_webhooks_controller.rb:30]: After a bit of investigation, it looks there's a _chance_ people can send attachments but we wouldn't be saving them at this time, so `has_documents` can be false

File: config/credentials/development.yml.enc (hunk 5 above)
Comment (jalcine-cfa) [reply config/credentials/development.yml.enc]: Done!

**Comment by jalcine-cfa** [comment]
Significant changes:

- Use `user` in Intercom as contact type more often
- Search for latest conversation in a different way
- Updated to use newish API methods based on API version 2.7
